### PR TITLE
Feature/api television

### LINKED ISSUE #411: Add video/TV natively in Gladys like the music abstraction
https://community.gladysproject.com/t/devicetype-de-type-push-button/2886/2

### PR DESCRIPTION
**Pull request by vincentBesseau**
Fixes #411
Clos #411
Topic on Gladys website :
https://community.gladysproject.com/t/integration-api-television-dans-gladys/3358

API definition :

gladys.television.pressKey()
gladys.television.getSources()
gladys.television.openSources()
gladys.television.openMenu()
gladys.television.rec()
gladys.television.play()
gladys.television.pause()
gladys.television.rewind()
gladys.television.fastForward()
gladys.television.switchState()
gladys.television.getState()
gladys.television.programPlus()
gladys.television.programMinus()
gladys.television.setChannel()
gladys.television.volumeDown()
gladys.television.volumeUp()
gladys.television.setMuted()
gladys.television.yellowButton()
gladys.television.greenButton()
gladys.television.redButton()
gladys.television.blueButton()
Until 4 devicetypes can be used :

Power devicetype
Sound devicetype
Channel devicetype
Mute devicetype
New sentences & new answer can be used with this api :

Open Channel 3 on TV on living room
=> The channel was changed !
And new box was added by @piznel
unknown-4
unknown-5

Historique PR #417 close pas erreur

capture d ecran 2018-09-18 a 22 15 29

capture d ecran 2018-09-18 a 22 15 40

### CODE REVIEW

File: api/controllers/TelevisionController.js [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+     * @apiSuccess {array} All television source
+     */
+    getSources: function(req, res, next){
+        gladys.television.getSources(req.query)
+            .then((result) => res.json(result))
+            .catch(next);
+    },
+
+    /**
+     * @api {post} /television/opensources Open Sources
```
Comment (Pierre-Gilles) [inline api/controllers/TelevisionController.js]: POST `/television/opensources`

Please keep the route singular! 

(I know every routes in Gladys are not singular, but most are)

File: api/core/television/television.sendCommand.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+          params.deviceType = deviceType;
+        
+          // calling television module
+          return gladys.modules[deviceType.service].television[functionName](params);
+        });
+};
+
+function getDeviceType(params){
+    if(params.device) return gladys.utils.sqlUnique(queries.getDeviceTypeByDeviceIdOrByDeviceTypeId, [params.device, params.device, null, null]);
```
Comment (Pierre-Gilles) [inline api/core/television/television.sendCommand.js]: I repost my comment here, why do you look for a deviceType/device when the key is device ? it's confusing, the API should not allow that!

File: api/core/television/television.sendCommand.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    else if(params.room) return gladys.utils.sqlUnique(queries.getTelevisionDeviceTypeByRoom, [params.room]);
+    else if(params.deviceType) return gladys.utils.sqlUnique(queries.getDeviceTypeByDeviceIdOrByDeviceTypeId, [null, null, params.deviceType, params.deviceType]);
```
Comment (Pierre-Gilles) [inline api/core/television/television.sendCommand.js]: Same here, I don't understand why you pass a deviceType and look for a deviceId or deviceTypeId, it's confusing!  

(btw, keep `devicetype` in lowercase as it's the case in the music API )

File: api/core/television/television.sendCommand.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+module.exports = function sendCommand(functionName, params) {
+    return getDeviceType(params)
+      .then((deviceType) => {
+
+          if(!gladys.modules[deviceType.service] || !gladys.modules[deviceType.service].television)
+                return Promise.reject(new Error(`television : Module ${deviceType.service} does not exist or does not handle television.`));
+         
+          if(typeof gladys.modules[deviceType.service].television[functionName] != 'function') {
+            console.log(`television : Module ${deviceType.service} does not have function ${functionName}`)
```
Comment (Pierre-Gilles) [inline api/core/television/television.sendCommand.js]: Use sails.log.error function

File: api/core/television/television.sendCommand.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+            return Promise.reject();
```
Comment (Pierre-Gilles) [inline api/core/television/television.sendCommand.js]: reject with a new Error(`television : Module ${deviceType.service} does not have function ${functionName}`);

File: api/core/television/television.customCommand.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+var sendCommand = require('./television.sendCommand.js');
+const Promise = require('bluebird');
+
+module.exports = function customCommand(params) {
+    if(params.color === 'red') return sendCommand('redButton', params);
+    else if(params.color === 'green') return sendCommand('greenButton', params);
+    else if(params.color === 'yellow') return sendCommand('yellowButton', params);
+    else if(params.color === 'blue') return sendCommand('blueButton', params);
+    else console.log('television : Button '+params.color+' not exist'); return Promise.reject();
```
Comment (Pierre-Gilles) [inline api/core/television/television.customCommand.js]: No console.log, use sails function

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
A few new feedbacks :)

File: api/controllers/TelevisionController.js (hunk 1 above)
Comment (vincentBesseau) [reply api/controllers/TelevisionController.js]: Fix this on commit 3ae5f94

File: api/core/television/television.sendCommand.js (hunk 4 above)
Comment (vincentBesseau) [reply api/core/television/television.sendCommand.js]: fix this on commit d86b44a

File: api/core/television/television.customCommand.js (hunk 6 above)
Comment (vincentBesseau) [reply api/core/television/television.customCommand.js]: fix this on commit d86b44a

File: api/core/television/television.sendCommand.js (hunk 3 above)
Comment (vincentBesseau) [reply api/core/television/television.sendCommand.js]: Fix this on commit 7088590

File: api/core/television/television.sendCommand.js (hunk 2 above)
Comment (vincentBesseau) [reply api/core/television/television.sendCommand.js]: Fix this on commit 7088590
Use only device to search the service

File: api/core/television/television.sendCommand.js (hunk 5 above)
Comment (vincentBesseau) [reply api/core/television/television.sendCommand.js]: Fix this on commit d593238

**Comment by vincentBesseau** [comment]
All feedback has been processed thank you @Pierre-Gilles.
For your point between the "device" and "devicetype" parameter, I would like to have a harmony between Music and television api.
After discuss with @piznel and @MathieuAndrade, we decided to keep only the "device" parameter to avoid confusion.

File: api/core/television/television.sendCommand.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+          params.deviceType = deviceType;
+        
+          // calling television module
+          return gladys.modules[deviceType.service].television[functionName](params);
+        });
+};
+
+function getDeviceType(params){
+    if(params.device.id) return gladys.utils.sqlUnique(queries.getDeviceTypeByDeviceId, [params.device.id]);
```
Comment (Pierre-Gilles) [inline api/core/television/television.sendCommand.js]: Just one latest feedback on this! 

In the music API, when we can the service, we do: 

```
gladys.music.play({ devicetype: 1 });
```

Here, with your implementation, we should do : 

```
gladys.music.play({ device: { id: 1} });
```

Not really consistent! 

Can we move to :

```
gladys.music.play({ device: 1 });
```
 ?

Thanks!

File: assets/js/app/television/television.controller.js [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+            vm.currentPowerState = null;
+            vm.currentSoundState = null;
+            vm.currentChannel = null;
+            vm.currentMuteState = null;
+            getData(deviceId);
+            getSources();
+        }
+
+        function getDevices() {
+            deviceService.getDeviceTypeByCategory({category:'TV'})
```
Comment (Pierre-Gilles) [inline assets/js/app/television/television.controller.js]: Just so you know, as here the feature is now called "television", I'm going to rename the category to "television" as well so it's more consistent.

I'll change the get here while merging! Any other place to change that?

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
I just tested all that on my laptop, and thanks guy, you did a really great work!! :) 

I added two little feedback, there is only one to take action on. 

Please answer on that, make the little change if you're good with it, and I'll merge! :)

File: api/core/television/television.sendCommand.js (hunk 7 above)
Comment (vincentBesseau) [reply api/core/television/television.sendCommand.js]: Changing on commit a4f0173

File: assets/js/app/television/television.controller.js (hunk 8 above)
Comment (vincentBesseau) [reply assets/js/app/television/television.controller.js]: Changing on commit 6f08686

**Comment by vincentBesseau** [comment]
Thank you @Pierre-Gilles :)
I fixed this two points on these commits a4f0173 & 6f08686.
I would like to thank @piznel & @MathieuAndrade for their helps and their works.

**Comment by Pierre-Gilles** [comment]
Hey! So tomorrow I'm morning I'm doing a review time of all PR in order to merge all pending PR, and I'll do a new Gladys version just after :)

I'm doing that now because I want to do a version before the Gladys Gateway version. 

I wanted to merge this (the feature is awesome, really good job again 👏 ), but I just realised that their are no tests in this PR!! 😢

I don't know if you're familiar with tests in Gladys, but for this kind of core native API, tests are just mandatory to ensure the API stability of this feature in time :) 

 Sorry for noticing this just now... :/ 

Still, I don't think it's a hard work, in 1 hour it could be done. I can help if needed :)

**Comment by Pierre-Gilles** [comment]
Thanks for this PR, just merged it! 👏

**Comment by Pierre-Gilles** [comment]
Just some little feedbacks, two things that could be corrected :) 

- The UI doesn't throw any error in case the requests are failing. Could it be possible to add error display in UI?

- The responsive is not that great =>

screenshot 2018-10-12 at 13 11 10

(Otherwise, I already merged the PR, it's already great like that!)

**Pull request closed without merging**